### PR TITLE
keys: do not use zip_iterator for printing key components

### LINKED ISSUE #14435: sstables/metadata_collector.cc:49 @update_min_max_components: failed to log message: fmt='{}: setting max_clustering_key={}': seastar::internal::backtraced<marshal_exception> (marshaling error: read_simple - not enough bytes (expected 4, got 0)
After enabling TestCommitLog in gating:
scylladb/scylla-dtest@e0cd2ac306deec541b17bc9ef5404cb2703ab3af

We immediately observed a failure
```
failed on teardown with "AssertionError: Unexpected errors found: [('node1', ["ERROR 2023-06-28 09:35:07,721 [shard 0] metadata_collector - sstables/metadata_collector.cc:49 @update_min_max_components: failed to log message: fmt='{}: setting max_clustering_key={}': seastar::internal::backtraced<marshal_exception> (marshaling error: read_simple - not enough bytes (expected 4, got 0) Backtrace: 0x5a7fbde 0x5a80190 0x5a80478 0x2d19948 0x2d19504 0x2cbd2dd 0x1c8c163 0x1c8b61d 0x1c8b574 0x1c8b129 0x1253d9c 0x25b9975 0x5a89d43 0x25b8ff6 0x23667e6 0x21ea871 0x21e9c06 0x59618a6", "ERROR 2023-06-28 09:35:07,720 [shard 0] metadata_collector - sstables/metadata_collector.cc:44 @update_min_max_components: failed to log message: fmt='{}: setting min_clustering_key={}': seastar::internal::backtraced<marshal_exception> (marshaling error: read_simple - not enough bytes (expected 4, got 0) Backtrace: 0x5a7fbde 0x5a80190 0x5a80478 0x2d19948 0x2d19504 0x2cbd2dd 0x1c8c163 0x1c8b61d 0x1c8b574 0x1c8b129 0x1253d9c 0x25b9975 0x5a89d43 0x25b9056 0x23667a5 0x21ea871 0x21e9c06 0x59618a6", "ERROR 2023-06-28 09:35:07,721 [shard 0] metadata_collector - sstables/metadata_collector.cc:49 @update_min_max_components: failed to log message: fmt='{}: setting max_clustering_key={}': seastar::internal::backtraced<marshal_exception> (marshaling error: read_simple - not enough bytes (expected 4, got 0) Backtrace: 0x5a7fbde 0x5a80190 0x5a80478 0x2d19948 0x2d19504 0x2cbd2dd 0x1c8c163 0x1c8b61d 0x1c8b574 0x1c8b129 0x1253d9c 0x25b9975 0x5a89d43 0x25b8ff6 0x23667a5 0x21ea871 0x21e9c06 0x59618a6", "ERROR 2023-06-28 09:35:07,725 [shard 0] metadata_collector - sstables/metadata_collector.cc:22 @convert: failed to log message: fmt='{}: convert: {}': seastar::internal::backtraced<marshal_exception> (marshaling error: read_simple - not enough bytes (expected 4, got 0) Backtrace: 0x5a7fbde 0x5a80190 0x5a80478 0x2d19948 0x2d19504 0x2cbd2dd 0x1c8c163 0x1c8b61d 0x1c8b574 0x1c8b129 0x1253d9c 0x25b984d 0x5a89d43 0x25b7b79 0x208b911 0x237590b 0x21e9c0e 0x59618a6", "ERROR 2023-06-28 09:35:07,725 [shard 0] metadata_collector - sstables/metadata_collector.cc:22 @convert: failed to log message: fmt='{}: convert: {}': seastar::internal::backtraced<marshal_exception> (marshaling error: read_simple - not enough bytes (expected 4, got 0) Backtrace: 0x5a7fbde 0x5a80190 0x5a80478 0x2d19948 0x2d19504 0x2cbd2dd 0x1c8c163 0x1c8b61d 0x1c8b574 0x1c8b129 0x1253d9c 0x25b984d 0x5a89d43 0x25b7b79 0x208b8fa 0x237590b 0x21e9c0e 0x59618a6"])]"
```

one of the lines where this is triggered:
```cpp
        mdclogger.trace("{}: setting min_clustering_key={}", _name, position_in_partition_view::printer(_schema, min_pos));
```

looks like `position_in_partition_view::printer(_schema, min_pos)` is incorrect - `min_pos` can't be printed in some cases?

### PR DESCRIPTION
**Pull request by tchaikov**
boost's the operator==() implementation of boost's zip_iterator returns true only if all elements in enclosed tuple of zip_iterator are equal. and the zip_iterator always advances all the iterators in the enclosed tuple. but in our case, some components might be missing. in other words, the size of the `components` might be smaller than that of the `types` range. so, when the zip_iterator advances past the end of the components, scylla starts reading out of bounds.

because zip_iterator does not allow us to customize how it implements the equal operator. and we cannot deduce the size of components without reading all of them. so in this change, we partially revert 3738fcbe054ef1540f262079d89e44c9b873769e, instead of using fmt::join(), just iterate through the components manually. this should avoid the out-of-bound reading, and also preserve the original behavior.

Fixes #14435
Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
@michoecho hi Michał, would you please help review this change?

File: keys.hh [hunk 1]
```diff
@@ -773,26 +770,23 @@ namespace detail {
...
+    for (auto&& component : key.components(schema)) {
+        if (!first) {
+            out = fmt::format_to(out, "{}", ":");
+        }
+        first = false;
+        auto key = (*type_iterator++)->to_string(to_bytes(component));
+        if (utils::utf8::validate((const uint8_t *) key.data(), key.size())) {
+            out = fmt::format_to(out, "{}", key);
+        } else {
+            out = fmt::format_to(out, "{}", "<non-utf8-key>");
```
Comment (michoecho) [inline keys.hh:785]: Offtop: we should probably print the non-utf8 key as a hex string or something, otherwise non-utf8 keys are not debuggable.

**Review by michoecho — APPROVED** [review]

**Comment by denesb** [comment]
I just fixed this in https://github.com/scylladb/scylladb/pull/14347 too (first commit).

File: keys.hh (hunk 1 above)
Comment (tchaikov) [reply keys.hh:785]: will create a pr to print them as hex.

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hi maintainers, could you help queue this change?

File: keys.hh (hunk 1 above)
Comment (tchaikov) [reply keys.hh:785]: #14498

**Pull request closed without merging**